service/common/src/main/java/org/apache/polaris/service/events/aws/cloudwatch/AwsCloudWatchEventListener.java

```diff
@@ -0,0 +1,275 @@
+        PutLogEventsRequest.builder()
+            .logGroupName(logGroup)
+            .logStreamName(logStream)
+            .logEvents(List.of(event));
+
+    synchronized (this) {
+      if (sequenceToken != null) {
+        requestBuilder.sequenceToken(sequenceToken);
+      }
+
+      try {
+        executePutLogEvents(requestBuilder);
+      } catch (InvalidSequenceTokenException e) {
+        sequenceToken = getSequenceToken();
+        requestBuilder.sequenceToken(sequenceToken);
+        executePutLogEvents(requestBuilder);
+      }
+    }
+  }
+
+  private void executePutLogEvents(PutLogEventsRequest.Builder requestBuilder) {
+    PutLogEventsResponse response = client.putLogEvents(requestBuilder.build());
+    sequenceToken = response.nextSequenceToken();
+  }
+
+  private void reapFuturesMap() {
+    for (Future<?> future : futures.keySet()) {
+      if (future.isDone()) {
+        Future.State currFutureState = future.state();
+        EventWithRetry currValue = futures.remove(future);
+        if (currFutureState.equals(Future.State.FAILED)
+            || currFutureState.equals(Future.State.CANCELLED)) {
+          if (currValue.retryCount >= 3) {
+            LOGGER.error("Event retries failed. Event dropped: {}", currValue.inputLogEvent);
+          } else {
+            EventWithRetry newValue =
+                new EventWithRetry(currValue.inputLogEvent, currValue.retryCount + 1);
+            future =
+                executorService.submit(() -> sendAndHandleCloudWatchCall(newValue.inputLogEvent));
+            futures.put(future, newValue);
+          }
+        }
+      }
+    }
+  }
+
+  private void transformAndSendEvent(HashMap<String, Object> properties) {
+    properties.put("realm", callContext.getRealmContext().getRealmIdentifier());
+    properties.put("principal", securityContext.getUserPrincipal().getName());
+    // TODO: Add request ID when it is available
+    String eventAsJson;
+    try {
+      eventAsJson = objectMapper.writeValueAsString(properties);
+    } catch (JsonProcessingException e) {
+      LOGGER.error("Error processing event into JSON string: {}", e.getMessage());
+      return;
+    }
+    InputLogEvent inputLogEvent = createLogEvent(eventAsJson, getCurrentTimestamp(callContext));
+    Future<?> future = executorService.submit(() -> sendAndHandleCloudWatchCall(inputLogEvent));
+    futures.put(future, new EventWithRetry(inputLogEvent, 0));
+  }
+
+  private long getCurrentTimestamp(CallContext callContext) {
+    return callContext.getPolarisCallContext().getClock().millis();
+  }
+
+  private InputLogEvent createLogEvent(String eventAsJson, long timestamp) {
+    return InputLogEvent.builder().message(eventAsJson).timestamp(timestamp).build();
+  }
+
+  @VisibleForTesting
+  ConcurrentHashMap<Future<?>, EventWithRetry> getFutures() {
+    return futures;
+  }
+
+  // Event overrides below
+  @Override
+  public void onAfterTableRefreshed(AfterTableRefreshedEvent event) {
```

Comment:
I still think we need to refactor these out so that the listener itself with all its Future-wrangling logic does not become exceedingly long